morph_tool/resampling.py

```diff
@@ -0,0 +1,201 @@
+    ids = np.searchsorted(path_lengths, new_path_lengths) - 1
+
+    # each fraction t corresponds to the fraction of the vector
+    # in the parametric equation of the line segment p0 + t * (p1 - p0)
+    fractions = (new_path_lengths - path_lengths[ids]) / (path_lengths[ids + 1] - path_lengths[ids])
+
+    return ids, fractions
+
+
+def _parametric_values(values, ids, fractions):
+    """It creates a new array of values, where the start and end
+    values remain unchanged and the in-between values are calculated
+    from the parametric equation:
+
+    v'[k] = v[ids[k]] + fractions[k] * (v[ids[k] + 1] - v[ids[k]])
+
+    Args:
+        values (np.ndarray): 1D or 2D array of values
+        ids (np.ndarray): (N,) Array of ints
+        fractions (np.ndarray): (N,) Array of floats
+
+    Returns:
+        new_values (np.ndarray): 1D or 2D array of length N + 2
+    """
+    n_new_values = len(ids) + 2
+
+    if values.ndim > 1:
+        fractions = fractions[:, np.newaxis]
+        new_values = np.empty_like(
+            values,
+            shape=(n_new_values, values.shape[1])
+        )
+    else:
+        new_values = np.empty_like(
+            values,
+            shape=(n_new_values,)
+        )
+
+    new_values[0] = values[0]
+    new_values[-1] = values[-1]
+    new_values[1:-1] = values[ids] + fractions * (values[ids + 1] - values[ids])
+
+    return new_values
+
+
+def _resample_section_neuron(section, linear_density):
+    """Resample in-place the section data based on linear density.
+    Args:
+        section (Section): Mutable morphology's section
+        linear_density (float): Linear density to determine the point number
+    """
+    points = section.points
+    ids, fractions = _resample_from_linear_density(points, linear_density)
+    section.points = _parametric_values(points, ids, fractions)
+    section.diameters = _parametric_values(section.diameters, ids, fractions)
+
+
+def _resample_section_astrocyte(section, linear_density):
+    """Resample in-place the section data based on linear density.
+    Args:
+        section (Section): Mutable morphology's section
+        linear_density (float): Linear density to determine the point number
+    """
+    points = section.points
+    ids, fractions = _resample_from_linear_density(points, linear_density)
+    section.points = _parametric_values(points, ids, fractions)
+    section.diameters = _parametric_values(section.diameters, ids, fractions)
+    section.perimeters = _parametric_values(section.perimeters, ids, fractions)
+
+
+def _dispatch_section_function(cell_family):
+    """Returns section function depending on the cell_family
+    of the morphology
+    """
+    return {
+        morphio.CellFamily.FAMILY_NEURON: _resample_section_neuron,
+        morphio.CellFamily.FAMILY_GLIA: _resample_section_astrocyte
+    }[cell_family]
+
+
+def resample_linear_density(obj, linear_density):
+    """Resample the number of points in morphology
```

Comment:
It looks like a good place, with the very top of this module, to explain what you mean by re-sampling (and why it might be required).

Reply:
Is it better now?

Reply:
yes